tests/unit/test_charm.py

```diff
@@ -0,0 +1,334 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Test cases for GithubRunnerCharm."""
+
+import os
+import unittest
+import urllib.error
+from unittest.mock import MagicMock, call, patch
+
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
+from ops.testing import Harness
+
+from charm import GithubRunnerCharm
+from errors import RunnerError, SubprocessError
+from github_type import GitHubRunnerStatus
+from runner_manager import RunnerInfo, RunnerManagerConfig
+from runner_type import GitHubOrg, GitHubRepo, VirtualMachineResources
+
+
+def raise_error(*args, **kargs):
+    raise Exception("mock error")
+
+
+def raise_runner_error(*args, **kargs):
+    raise RunnerError("mock error")
+
+
+def raise_subprocess_error(*args, **kargs):
+    raise SubprocessError(cmd=["mock"], return_code=1, stdout="mock stdout", stderr="mock stderr")
+
+
+def raise_url_error(*args, **kargs):
+    raise urllib.error.URLError("mock error")
+
+
+def mock_get_latest_runner_bin_url():
+    mock = MagicMock()
+    mock.download_url = "www.example.com"
+    return mock
+
+
+def mock_get_github_info():
+    return [
+        RunnerInfo("test runner 0", GitHubRunnerStatus.ONLINE.value),
+        RunnerInfo("test runner 1", GitHubRunnerStatus.ONLINE.value),
+        RunnerInfo("test runner 2", GitHubRunnerStatus.OFFLINE.value),
+        RunnerInfo("test runner 3", GitHubRunnerStatus.OFFLINE.value),
+        RunnerInfo("test runner 4", "unknown"),
+    ]
+
+
+class TestCharm(unittest.TestCase):
+    """Test cases for GithubRunnerCharm."""
+
+    @patch.dict(
+        os.environ,
+        {
+            "JUJU_CHARM_HTTPS_PROXY": "mock_https_proxy",
+            "JUJU_CHARM_HTTP_PROXY": "mock_http_proxy",
+            "JUJU_CHARM_NO_PROXY": "mock_no_proxy",
+        },
+    )
+    def test_proxy_setting(self):
+        harness = Harness(GithubRunnerCharm)
+        harness.begin()
+
+        assert harness.charm.proxies["https"] == "mock_https_proxy"
+        assert harness.charm.proxies["http"] == "mock_http_proxy"
+        assert harness.charm.proxies["no_proxy"] == "mock_no_proxy"
+
+    @patch("pathlib.Path.write_text")
+    @patch("subprocess.run")
+    def test_install(self, run, wt):
```

Comment:
For all the tests: I missed the docstrings as described in the [contributing guide](https://github.com/canonical/is-charms-contributing-guide/blob/main/CONTRIBUTING.md#test-structure).

Reply:
Same as above